### PR TITLE
[Issue 12962] Fix should not check if partition topic here.

### PR DESCRIPTION
**Pull request by casuallc**
### Motivation

If we consume more then one topics, can not create TopicMessageId for non-partition topic, so `checkArgument(messageId instanceof TopicMessageIdImpl);` would faild.

```
    @Override
    protected CompletableFuture<Void> doAcknowledge(MessageId messageId, AckType ackType,
                                                    Map<String,Long> properties,
                                                    TransactionImpl txnImpl) {
        checkArgument(messageId instanceof TopicMessageIdImpl);
        TopicMessageIdImpl topicMessageId = (TopicMessageIdImpl) messageId;

        if (getState() != State.Ready) {
            return FutureUtil.failedFuture(new PulsarClientException("Consumer already closed"));
        }
        .....
  }
```

#12962 

.

### Modifications
Remove check partition.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 
  
- [x] `no-need-doc`

### CODE REVIEW

**Review by codelipenghui — COMMENTED** [review]
Nice catch! Could you please add a test to avoid the regression?

**Comment by casuallc** [comment]
@codelipenghui  Test added, PTAL.

**Review by codelipenghui — DISMISSED** [review]

**Comment by codelipenghui** [comment]
Sorry @casuallc, we already have a test class `MessageIdSerializationTest.java` there, could you please move the test to `MessageIdSerializationTest.java`

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/MessageIdTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import org.apache.pulsar.client.api.MessageId;
+import org.testng.annotations.Test;
+
+public class MessageIdTest {
```
Comment (nodece) [inline pulsar-client/src/test/java/org/apache/pulsar/client/impl/MessageIdTest.java:24]: I think you don't need to create a new test file, you should add the test case in [here](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/MessageIdSerializationTest.java).

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java [hunk 2]
```diff
@@ -159,7 +159,7 @@ public static MessageId fromByteArrayWithTopic(byte[] data, TopicName topicName)
         } else {
             messageId = new MessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition());
         }
-        if (idData.getPartition() > -1 && topicName != null) {
+        if (topicName != null) {
```
[the remark this answers is not in the document]
Comment (casuallc) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:162]: Consumer should tell broker which topic message to be acknowledged.

`MultiTopicsConsumerImpl` would check if `MessageId` is `TopicMessageIdImpl` on method `doAcknowledge`.
```
@Override
protected CompletableFuture<Void> doAcknowledge(MessageId messageId, AckType ackType,
                                                Map<String,Long> properties,
                                                TransactionImpl txnImpl) {
    checkArgument(messageId instanceof TopicMessageIdImpl);
    TopicMessageIdImpl topicMessageId = (TopicMessageIdImpl) messageId;
    // ....
}
```

[the remark this answers is not in the document]
Comment (315157973) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:162]: Only the type of the message ID of the partitioned topic is TopicMessageIdImpl, you will also change the message ID of the non-partitioned topic to TopicMessageIdImpl

[the remark this answers is not in the document]
Comment (casuallc) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:162]: I do not get you point...  This is my test code.

Env: pulsar-2.8.0
```
List<String> topics = new ArrayList<>();
topics.add("topic01");
topics.add("topic02");

final Consumer<String> consumer1 = client.newConsumer(Schema.STRING)
		.consumerName("vv-01")
		.topics(topics)
		.subscriptionName("sub-01")
		.subscriptionType(SubscriptionType.Shared)
		.subscribe();

System.out.println(consumer1.getClass()); // org.apache.pulsar.client.impl.MultiTopicsConsumerImpl

Message<String> msg = consumer1.receive();
System.out.println(msg.getMessageId()); // 10:1:-1:0
System.out.println(msg.getMessageId().getClass()); // class org.apache.pulsar.client.impl.TopicMessageIdImpl
consumer1.acknowledge(msg);

consumer1.close();
```

I think both partitioned and non-partitioned topic need `TopicMessageIdImpl`.

[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:162]: Sorry, Looks like I missed something here. If the consumer only subscribes to a single non-partitioned topic, we should return the MessageIdImpl, not the TopicMessageIdImpl right?

[the remark this answers is not in the document]
Comment (casuallc) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:162]: Sorry, I know what you mean, but the API is confused...

image

[the remark this answers is not in the document]
Comment (nodece) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java]: Firstly commit looks good to me, the `fromByteArrayWithTopic()` should return a `TopicMessageIdImpl`. Could we add an integration test to test this?

**Comment by BewareMyPower** [comment]
Could you continue this work?

**Comment by congbobo184** [comment]
@casuallc   hi, I move this PR to release/2.9.5, if you have any questions, please ping me. thanks.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

I believe this patch is valuable to continue. But keeping a stale conflict PR open doesn't help. Rebase and resubmit is always available and welcome. @BewareMyPower you may integrate this work during your effort on the message id refactor. Keep @casuallc in the "Co-authored-by" footer.

**Comment by BewareMyPower** [comment]
@tisonkun I don't think we should continue this work. Adding a new API into the `pulsar-client` module is bad. We should not encourage application users to use the APIs in this module. If they need, we should open a PIP to add a new API into the `pulsar-client-api` module.

**Pull request closed without merging**